### PR TITLE
K.F 第四次作业

### PR DESCRIPTION
**Pull request by KFGameHacker**

### CODE REVIEW

**Comment by KFGameHacker** [comment]
@kaichaosun

File: projects/lesson-3/runtime/src/kitties.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+		/// Create a new kitty
+		pub fn create(origin) {
+			let sender = ensure_signed(origin)?;
+			let count = Self::kitties_count();
+			if count == u32::max_value() {
+				return Err("Kitties count overflow");
+			}
+			let payload = (<system::Module<T>>::random_seed(), sender, <system::Module<T>>::extrinsic_index(), <system::Module<T>>::block_number());
+			let dna = payload.using_encoded(blake2_128);
+			let kitty = Kitty(dna);
```
Comment (kaichaosun) [inline projects/lesson-3/runtime/src/kitties.rs]: You are using another Kitty struct.

File: projects/lesson-3/runtime/src/kitties.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+			Kitties::insert(count, kitty);
+			KittiesCount::put(count + 1);
+		}
+
+		//breed a kitty from parent
+		pub fn breed_kitty(origin,father_kitty_hash: T::Hash, mother_kitty_hash: T::Hash){
```
Comment (kaichaosun) [inline projects/lesson-3/runtime/src/kitties.rs]: Good naming, but you probably mean index not hash.

File: projects/lesson-3/runtime/src/kitties.rs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+			// parents hash must exist
+			ensure!(<Kittyies<T>>::exists(father_kitty_hash));
+			ensure!(<Kittyies<T>>::exists(mother_kitty_hash));
+
+			// get parents struct
+			let father_kitty = Self::kitty(father_kitty_hash);
+			let mother_kitty_hash = Self::kitty(father_kitty_hash);
+
+			// todo: mix parents DNA to generate child new DNA
+			let child_dna = 0;
```
Comment (kaichaosun) [inline projects/lesson-3/runtime/src/kitties.rs]: The dna should be random and inheritant from parents.

File: projects/lesson-3/runtime/src/kitties.rs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+                price: <T::Balance as As<u64>>::sa(0),
+                gen: cmp::max(kitty_1.gen, kitty_2.gen) + 1,
+            };
+
+			// checking before inserting 
+			ensure!(!<Kitties<T>>::exists(child_kitty.id), "Kitty already exists");
+
+			// add the new child kitty to the data storage
+			Kitties::insert(count, kitty);
+			KittiesCount::put(count + 1);
```
Comment (kaichaosun) [inline projects/lesson-3/runtime/src/kitties.rs]: Need to check the overflow before add.

File: projects/lesson-4/runtime/src/kitties.rs [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+		}
+
+		/// Breed kitties
+		pub fn breed(origin, kitty_id_1: T::KittyIndex, kitty_id_2: T::KittyIndex) {
+			let sender = ensure_signed(origin)?;
+
+			Self::do_breed(sender, kitty_id_1, kitty_id_2)?;
+		}
+
+		pub fn transfer_kitty(origin, T::Account, kitty_id: T::KittyIndex) -> Result {
```
Comment (kaichaosun) [inline projects/lesson-4/runtime/src/kitties.rs]: The signature is not correct, should be `to: T::Account`

File: projects/lesson-4/runtime/src/kitties.rs [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+			.ok_or("transfer error of 'from' account.");
+
+		// safe add one for 'to' account owned index
+		let new_to_account_owned_kitties = from_account_owned_kitties.checked_add(1)
+			.ok_or("transfer error of 'to' account.");
+
+		// get kitty index 
+		let kitty_index = <OwnedKittiesIndex<T>>::get(kitty_id);
+
+		if kitty_index != new_from_account_owned_kitties {
```
Comment (kaichaosun) [inline projects/lesson-4/runtime/src/kitties.rs]: Good to see you are use swap and upate.

**Pull request merged**